relayer/crates/starknet-chain-components/src/types/channel_id.rs

```diff
@@ -1,220 +1,139 @@
-            Product![
-                ChannelState,
-                ChannelOrdering,
-                ChannelCounterparty,
-                ConnectionId,
-                AppVersion,
-            ],
-        > + CanRaiseAsyncError<&'static str>,
+    Encoding: CanEncodeMut<Strategy, RawChannelEnd> + CanRaiseAsyncError<&'static str>,
 {
     fn encode_mut(
         encoding: &Encoding,
         value: &ChannelEnd,
         buffer: &mut Encoding::EncodeBuffer,
     ) -> Result<(), Encoding::Error> {
-        if value.connection_hops.len() != 1 {
-            return Err(Encoding::raise_error("invalid connection hops"));
-        }
-
-        encoding.encode_mut(
-            &product![
-                value.state,
-                value.ordering,
-                value.counterparty().clone(),
-                value.connection_hops[0].clone(),
-                value.version.clone(),
-            ],
-            buffer,
-        )?;
-        Ok(())
+        let counterparty = value.counterparty();
+
+        let channel_id = counterparty
+            .channel_id
+            .as_ref()
+            .map(ToString::to_string)
+            .unwrap_or_else(String::new);
+
+        let [connection_id]: [ConnectionId; 1] = value
+            .connection_hops
+            .clone()
+            .try_into()
+            .map_err(|_| Encoding::raise_error("expect exactly one connection id to exist"))?;
+
+        let raw = RawChannelEnd {
+            state: (&value.state).into(),
+            ordering: value.ordering,
+            remote: RawChannelCounterparty {
+                port_id: counterparty.port_id.clone(),
+                channel_id,
+            },
+            connection_id,
+            version: value.version.clone(),
+        };
+
+        encoding.encode_mut(&raw, buffer)
     }
 }
 
 #[cgp_provider(MutDecoderComponent)]
 impl<Encoding, Strategy> MutDecoder<Encoding, Strategy, ChannelEnd> for EncodeChannelEnd
 where
-    Encoding: CanDecodeMut<
-            Strategy,
-            Product![
-                ChannelState,
-                ChannelOrdering,
-                ChannelCounterparty,
-                ConnectionId,
-                AppVersion,
-            ],
-        > + CanRaiseAsyncError<&'static str>,
+    Encoding: CanDecodeMut<Strategy, RawChannelEnd>
+        + CanRaiseAsyncError<ChannelError>
+        + CanRaiseAsyncError<IdentifierError>,
 {
     fn decode_mut<'a>(
         encoding: &Encoding,
         buffer: &mut Encoding::DecodeBuffer<'a>,
     ) -> Result<ChannelEnd, Encoding::Error> {
-        let product![state, ordering, counterparty, connection_id, version,] =
-            encoding.decode_mut(buffer)?;
+        let raw = encoding.decode_mut(buffer)?;
 
-        ChannelEnd::new(state, ordering, counterparty, vec![connection_id], version)
-            .map_err(|_| Encoding::raise_error("invalid channel end"))
+        let raw_channel_id = raw.remote.channel_id;
+        let channel_id = if raw_channel_id.is_empty() {
```

Comment:
We probably need to fix the Cairo contract to use `Option` for optional channel ID, instead of relying on the non-standard behavior of using empty string to represent non-existence.

Reply:
definitely. last time I checked with Farhad, the main issue that, Cairo uses default (empty) values when querying from storage. that's why it's implemented that way.

I just checked with the latest cairo, it seems, I can put `Option<T>` in `Storage` derived types. We should do it.